mlflow/data/dataframe_dataset.py

```diff
@@ -0,0 +1,265 @@
+from __future__ import annotations
+
+import json
+import logging
+from functools import cached_property
+from typing import Any, Generic, Optional, Union
+
+import narwhals.stable.v1 as nw
```

Comment:
For some context: by import narwhals through its stable api, we guarantee that all non-unstable features will not have breaking changes in the future. If we were to make backward incompatible changes in the main namespace, we would also take the effort to maintain v1 stable.

As an example, we removed `LazyFrame.gather_every` method in the main namespace (due to future no guarantee in row order), but kept it in the stable v1 version (https://github.com/narwhals-dev/narwhals/pull/2103).

Useful reading: [Perfect backwards compatibility policy
](https://narwhals-dev.github.io/narwhals/backcompat/)